sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -691,7 +692,8 @@ private[sql] class SQLConf extends Serializable with CatalystConf with Logging {
   def variableSubstituteDepth: Int = getConf(VARIABLE_SUBSTITUTE_DEPTH)
 
   def warehousePath: String = {
-    getConf(WAREHOUSE_PATH).replace("${system:user.dir}", System.getProperty("user.dir"))
+    new File(getConf(WAREHOUSE_PATH).replace("${system:user.dir}",
+      System.getProperty("user.dir"))).toURI.toString
```

Comment:
You are correct, `toURI` behaves strange when the path is not local:
```
scala> new File("hdfs://m.com:9000/data").toURI
res6: java.net.URI = file:/C:/Program%20Files%20(x86)/scala/bin/hdfs:/m.com:9000
/data
```
I think we can add a check and apply `toURI` when the string does not start from `hdfs:` or `s3`.